### DISCUSSION TITLE
Is interruption of `time.sleep` supported?

### DISCUSSION THREAD

**Discussion by akshayka** [pyodide/pyodide#4595] [Q&A]
I recently implemented support for interrupting Python execution using a `SharedArrayBuffer`, as described in the docs (I'm running Pyodide in a WebWorker). It works very well for the most part. For example, I can interrupt code like

```python
while True:
   ...
```

with no issue.

I did notice one peculiarity. I'm unable to interrupt the execution `time.sleep(...)`, for any number of seconds. Is this a known issue, or am do you suspect I'm doing something wrong?

**Comment by ryanking13** [comment]
IIRC `time.sleep()` was no-op in Pyodide, but interestingly it seems to work now. @hoodmane Do you remember when it became available? Maybe it is related to the [`time.sleep()` change in Python 3.11](https://docs.python.org/3/whatsnew/3.11.html#time)?

**Comment by hoodmane** [comment]
Thanks for the report @akshayka, this is a known bug. I'm not sure that anyone has complained about it before, but it annoys me.

@ryanking13 we picked up the busy-loop sleep from an Emscripten change, I believe some time after the Python 3.11 update. It is not related to any Python change. I have this PR to try to improve Emscripten's sleep:
https://github.com/emscripten-core/emscripten/pull/20011

I think the correct fix for this is to switch to an atomics-based sleep in a webworker and call `pyodide.checkInterrupt` periodically, say once every 50 ms:
```js
function sleep(ms) {
   for (let i = 0; i < ms/50; i++) {
      Atomics.wait(waitBuffer, 0, 0, 50);
      pyodide.checkInterrupt();
   }
   Atomics.wait(waitBuffer, 0, 0, ms % 50);
}
```
Of course we could also periodically call `pyodide.checkInterrupt` for the busy sleep, but if we can't block people are unlikely to be able to use interrupts either.

**Reply by ryanking13** [reply]
I see. Didn't know that Emscripten was using busy sleep internally.

**Reply by akshayka** [reply]
Thanks for explaining. It's not a huge issue; it was mostly surprising, as I was using `time.sleep()` as an ad-hoc test case of something that should be interruptible -- for a while I thought my implementation was incorrect.

**Reply by hoodmane** [reply]
There's a lot of minor warts like this, it's nice to get rid of them when we can.
